### ISSUE TITLE
Communication from world with lower MTU is failing to LoadBalancer services

### ISSUE THREAD

**Issue by nextsux** [cilium/cilium#34380]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

equal or higher than v1.16.0 and lower than v1.17.0

### What happened?

I'm having troubles talking onto LoadBalancer services from outside "world" when user is on lower MTU than cluster's.

It seems like ICMP "fragmentation needed" is not delivered to the pod. Instead it loops on k8s host until its TTL exceedes. 

I've seen lots of issues around this but it LOOKS like it SHOULD (https://github.com/cilium/cilium/issues/12854 https://github.com/industrialdiscipline/cilium/commit/07edf550cab7f865ef83522061256183a803109e) be solved. Unfortunately not for me. 

Interesting is I can't see any drops of ICMP using cilium monitor. TBH I'm pretty new to cilium so I'm a little confused how to debug this any further. So I've tried to use tcpdump all around and draw a scheme of what I think is happening where red line is what is **wrong** and dotted lines are what SHOULD happen. I hope it makes sense.

Cilium is installed with helm using these values:

### How can we reproduce the issue?

* Install cilium with helm using this values:

```
bpf:
  hostLegacyRouting: false
  masquerade: true
cluster:
  name: <CUT>
cni:
  customConf: false
  uninstall: false
ipam:
  mode: cluster-pool
  operator:
    clusterPoolIPv4PodCIDRList:
      - 10.98.0.0/16
    clusterPoolIPv6PodCIDRList:
      - fd11:1111:1111::/112
operator:
  replicas: 1
  unmanagedPodWatcher:
    restart: true
policyEnforcementMode: default
routingMode: tunnel
tunnelProtocol: vxlan
kubeProxyReplacement: true
k8sServiceHost: <CUT>
k8sServicePort: 6443
hubble:
  relay:
    enabled: false
  ui:
    <CUT>
ipv4:
  enabled: true

ipv6:
  enabled: true

pmtuDiscovery:
  enabled: true

fragmentTracking: true
```

* create a ciliumloadbalancerippools
* create a service of type LoadBalancer, let it allocate IP
* Have a machine with lower MTU somewhere else. If your cluster is using 1500, use for example 1280.
* Try to access service IP (which should return more data then MTU for issue to reproduce). For example use web server and return some html file bigger than 1500B)

### Cilium Version

v1.16.0

### Kernel Version

6.1.0-22-amd64

### Kubernetes Version

v1.30.2

### Sysdump

[cilium-sysdump-20240814-134245.zip](https://github.com/user-attachments/files/16612345/cilium-sysdump-20240814-134245.zip)

### Relevant log output

```shell
23:50:14.727684 enp1s0 Out ifindex 2 52:54:00:4f:ef:df ethertype IPv4 (0x0800), length 596: (tos 0xc0, ttl 51, id 35597, offset 0, flags [none], proto ICMP (1), length 576)
    88.146.175.155 > 192.168.123.11: ICMP 88.146.175.155 unreachable - need to frag (mtu 1350), length 556
	(tos 0x0, ttl 53, id 48450, offset 0, flags [DF], proto TCP (6), length 1450)
    192.168.123.11.443 > 88.146.175.155.50138: Flags [.], seq 0:1398, ack 1, win 501, options [nop,nop,TS val 1471989322 ecr 3720111057], length 1398
23:50:14.727725 enp1s0 In  ifindex 2 52:54:00:62:06:d4 ethertype IPv4 (0x0800), length 596: (tos 0xc0, ttl 50, id 35597, offset 0, flags [none], proto ICMP (1), length 576)
    88.146.175.155 > 192.168.123.11: ICMP 88.146.175.155 unreachable - need to frag (mtu 1350), length 556
	(tos 0x0, ttl 53, id 48450, offset 0, flags [DF], proto TCP (6), length 1450)
    192.168.123.11.443 > 88.146.175.155.50138: Flags [.], seq 0:1398, ack 1, win 501, options [nop,nop,TS val 1471989322 ecr 3720111057], length 1398
23:50:14.727734 enp1s0 Out ifindex 2 52:54:00:4f:ef:df ethertype IPv4 (0x0800), length 596: (tos 0xc0, ttl 49, id 35597, offset 0, flags [none], proto ICMP (1), length 576)
    88.146.175.155 > 192.168.123.11: ICMP 88.146.175.155 unreachable - need to frag (mtu 1350), length 556
	(tos 0x0, ttl 53, id 48450, offset 0, flags [DF], proto TCP (6), length 1450)
    192.168.123.11.443 > 88.146.175.155.50138: Flags [.], seq 0:1398, ack 1, win 501, options [nop,nop,TS val 1471989322 ecr 3720111057], length 1398
23:50:14.727752 enp1s0 In  ifindex 2 52:54:00:62:06:d4 ethertype IPv4 (0x0800), length 596: (tos 0xc0, ttl 48, id 35597, offset 0, flags [none], proto ICMP (1), length 576)
    88.146.175.155 > 192.168.123.11: ICMP 88.146.175.155 unreachable - need to frag (mtu 1350), length 556
	(tos 0x0, ttl 53, id 48450, offset 0, flags [DF], proto TCP (6), length 1450)
    192.168.123.11.443 > 88.146.175.155.50138: Flags [.], seq 0:1398, ack 1, win 501, options [nop,nop,TS val 1471989322 ecr 3720111057], length 1398
```

### Anything else?

![obrazek](https://github.com/user-attachments/assets/477a7f3c-f678-4096-b0c0-9e089849f612)

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by julianwiedmann** [comment]
>It seems like ICMP "fragmentation needed" is not delivered to the pod. Instead it loops on k8s host until its TTL exceedes.

Yep, that's currently expected behaviour. For this to work, the LB code would need to peek into the ICMP packet, understand that it's a service access, match it against the established connection towards the backend, and translate the outer & inner headers accordingly before delivering to the backend. I'm not aware that we have such logic in place today.

**Comment by haribhusal2025** [comment]
We are experiencing this issue as well, where if cluster nodes have 9000 mtu, communication between cluster and world is affected although cilium doesn't show any drops.

**Comment by maxpain** [comment]
I have a similar problem.
My cluster MTU is 9000, and I'm trying to use `NodePort` to expose a service to the Internet via the interface with MTU 1400.

The requests coming from the internet to pods are routed fine, but responses are dropped by Cilium with the `FIB lookup failed` reason.

**Comment by maxpain** [comment]
not stale

**Comment by nerdicbynature** [comment]
We have the same issue. Definitely not stale. How can we proceed here?

**Comment by veith4f** [comment]
=== TEST CURL PAKETSIZE > MTU 9000 // gardener // gardener cilium-extension:v1.36.0 ===

Wanted to test if this problem persists. Tested using 
- gardener (https://github.com/gardener/gardener) 
- with cilium extension (https://github.com/gardener/gardener-extension-networking-cilium)
- bitnami nginx (oci://registry-1.docker.io/bitnamicharts/nginx)
- on our gardener installation where mtu = 9000

DEFINITIONS
===
https://github.com/veith4f/mtutest
```bash
git clone https://github.com/veith4f/mtutest
ls -lah mtutest
total 60K
drwxr-xr-x  8 fschuetz staff  256 Feb 11 11:01 .
drwxr-xr-x  6 fschuetz staff  192 Feb 11 11:11 ..
drwxr-xr-x 13 fschuetz staff  416 Feb 11 11:02 .git
-rw-r--r--  1 fschuetz staff  26K Feb 11 10:54 LICENSE
-rw-r--r--  1 fschuetz staff    9 Feb 11 10:54 README.md
-rw-r--r--  1 fschuetz staff  615 Feb 11 10:54 index.html
-rw-------  1 fschuetz staff 8.3K Feb 11 11:00 logo.svg
-rw-r--r--  1 fschuetz staff 8.8K Feb 11 11:00 mtu.html 
```
values.yaml
```yaml
podAntiAffinityPreset: hard
replicaCount: 1 # set high enough to get a node logging a HealthCheck problem
service:
  externalTrafficPolicy: Local
cloneStaticSiteFromGit:
  enabled: true
  repository: https://github.com/veith4f/mtutest
  branch: main
```

TEST
===
 
1. check client mtu size: 
ifconfig en7
en7: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=6464<VLAN_MTU,TSO4,TSO6,CHANNEL_IO,PARTIAL_CSUM,ZEROINVERT_CSUM>
	ether c4:41:1e:68:d2:f7
	inet6 fe80::1414:6f37:7940:c619%en7 prefixlen 64 secured scopeid 0x10 
	inet 192.168.178.51 netmask 0xffffff00 broadcast 192.168.178.255
	inet6 2a05:fa40:106:e400:8cf:24a2:e44f:17f3 prefixlen 64 autoconf secured 
	inet6 2a05:fa40:106:e400:5dc2:e0fa:d797:8bec prefixlen 64 autoconf temporary 
	inet6 fd59:a539:3a69:0:cd8:ab78:e5b:784a prefixlen 64 autoconf secured 
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect (100baseTX <full-duplex>)
	status: active
2. check cilium extension version. 
kubectl describe deployment -n extension-networking-cilium-jqnbk gardener-extension-networking-cilium | grep Image
    Image:      europe-docker.pkg.dev/gardener-project/public/gardener/extensions/networking-cilium:v1.36.02.
3. create new gardener cluster with k8s 1.27
4. helm install nginx oci://registry-1.docker.io/bitnamicharts/nginx -f values.yaml
5. wait for external ip
6. kubectl get svc

NAME         TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)                      AGE

kubernetes   ClusterIP      10.112.0.1     <none>         443/TCP                      5m

nginx        LoadBalancer   10.119.95.76   78.138.72.56   80:30175/TCP,443:30609/TCP   61s

6. kubectl -n kube-system exec -ti ds/cilium – cat /tmp/cilium/config-map/enable-bpf-masquerade

Defaulted container "cilium-agent" out of: cilium-agent, disable-rp-filter (init), config (init), mount-cgroup (init), apply-sysctl-overwrites (init), mount-bpf-fs (init), clean-cilium-state (init), install-cni-binaries (init)

true%

7. curl 78.138.72.56/logo.svg
... works
8. curl 78.138.72.56/mtu.html
... works

9. increase size of mtu.html in mtutest repo: 

or i in $(seq 1 9000); do echo -n "f" >> mtu.html; done; git add mtu.html; git commit ...

10. helm uninstall nginx

11. helm install nginx oci://registry-1.docker.io/bitnamicharts/nginx -f values.yaml

12. kubectl get svc

NAME         TYPE           CLUSTER-IP       EXTERNAL-IP    PORT(S)                      AGE

kubernetes   ClusterIP      10.112.0.1       <none>         443/TCP                      20m

nginx        LoadBalancer   10.127.104.170   78.138.72.74   80:31146/TCP,443:32030/TCP   4m27s

13. curl 78.138.72.74/mtu.html > mtu.html

rw-rr- 1 fschuetz staff 18K Feb 11 11:51 mtu.html

RESULT
===
In this test it looks like communication from client with mtu = 1500 with service/pod on cluster with mtu = 9000 is working. Can anyone confirm?

**Comment by pasteley** [comment]
> Did I miss anything? 

As far as I understand, the problem occurs when there is middleware router that responses with ICMP Type 2 Kind (ICMPv6 Packet Too Big), that allows pmtu discovery to work. And looks like only way to handle this is to apply mss clamping (for tcp) on that middleware, or adding lower mtu route for backwards route.

**Comment by nextsux** [comment]
> only way to handle this

or to implement correct handling of fragmentation needed ICMPs

**Comment by veith4f** [comment]
> > Did I miss anything?
> 
> As far as I understand, the problem occurs when there is middleware router with outbound interface MTU being lower then inbound, that responses with ICMP Type 2 Kind (ICMPv6 Packet Too Big), which allows pmtu discovery to work. And looks like only way to handle this is to apply mss clamping (for tcp) on that middleware, or adding lower mtu route for backwards route.

Are we talking about the exact issue described in https://github.com/cilium/cilium/issues/29583?

Are you aware of steps to reproduce?

**Comment by newotherworld** [comment]
yep. not stale.

**Comment by veith4f** [comment]
Yep. Not stale.

**Comment by centromere** [comment]
Hello. Has anyone identified a solution to this issue?

**Comment by minhng99** [comment]
Hello, I'm also having the same issue here:
- My cluster is using MTU 9000 on all of its nodes, with default-gateway route having a custom 1500 MTU because I'm not expecting L3 to do MTU 9000
- Any incoming requests to `node-port` is correctly negotiated down to MSS 1460 in the SYN-ACK.
- Any incoming requests to `External IP` that advertised by BGP is still trying to do MSS 8910 in the SYN-ACK. Because of this, my services is unstable if someone calling it from outside the L2 which is not jumbo-frame capable.

**Comment by vjeko** [comment]
I have not done extensive analysis, but this issue is only seen when running in eBPF (masquerading) mode. Reason being, there is no way to instruct eBPF to perform MSS clamping on the interface with the lower MTU.

After switching to legacy routing/masquerading, I am no longer seeing this issue, since packets are now flowing through iptables/nftables where I can insert a clamping rule.

**Comment by joshgordon** [comment]
I bashed my head against the wall for a while on this today. It's not a perfect solution, as in it only works with TCP, but if you can get a router in the path somewhere to do a little mangle and set the MSS on the initial SYN, cilium seems to respect it when it starts sending traffic back. 

For instance, I have a cluster that has it's own router and a wireguard interface (MTU 1420 / MSS 1380) back to the rest of my network. I added this rule to my iptables: 

```
iptables -t mangle -A FORWARD -o wg0 -p tcp -m tcp --tcp-flags SYN,RST SYN -j TCPMSS --set-mss 1380
iptables -t mangle -A FORWARD -i wg0 -p tcp -m tcp --tcp-flags SYN,RST SYN -j TCPMSS --set-mss 1380
```

and now things seem to be working as expected.

**Comment by wind0r** [comment]
We also hit the issue that "fragmentation needed" ICMP packets aren't handled on our all-L3 BGP/ECMP clusters. We can't clamp the requests since it would cause too big of a performance hit. Clamping imo also isn't the correct solution — the ICMP packets just should be handled.

I've got a datapath fix in progress that handles the ICMP "fragmentation needed" / "packet too big" properly instead of relying on an upstream router to clamp — draft PR: https://github.com/cilium/cilium/pull/46993.

Short version of the approach: intercept the ICMP error addressed to the service VIP, parse the embedded original reply to identify the service and connection, and forward the error on to the backend that needs to lower its PMTU — re-deriving the owning backend statelessly so it works even when BGP/ECMP lands the error on a different node than the one holding the flow.

Currently just for Maglev, and not yet with SNAT or overlay networking.

**Comment by yushoyamaguchi** [comment]
@wind0r 

Thank you for your work!

I think I may be working on a similar problem, but taking a somewhat different approach. I'm currently working on handling RevDNAT for ICMP error packets related to service traffic.
https://github.com/cilium/cilium/pull/46939

I haven't compared the two approaches in detail yet, but they seem closely related, so I wanted to leave a comment here. I'll take a closer look at your approach and see how they overlap or differ.

**Comment by yushoyamaguchi** [comment]
@wind0r 

I looked more and I don't think our PRs overlap. Mine only handles errors sent by the backend itself, using conntrack. Yours handles errors from routers outside the cluster, addressed to the VIP, which can land on any node — mine can't do that. So they're different problems. 
I'm sorry for the confusion!
